crates/common/src/transaction.rs

```diff
@@ -207,3 +277,978 @@ pub enum DataAvailabilityMode {
+        }
+        .hash(chain_id)
+    }
+}
+
+#[derive(Default)]
+struct LegacyHasher {
+    pub prefix: Felt,
+    pub address: ContractAddress,
+    pub entry_point: EntryPoint,
+    pub data_hash: Felt,
+    pub nonce: Option<TransactionNonce>,
+}
+
+impl LegacyHasher {
+    fn hash(self, chain_id: ChainId) -> TransactionHash {
+        let mut hasher = PedersenHasher::default()
+            .chain_update(self.prefix)
+            .chain_update(*self.address.get())
+            .chain_update(self.entry_point.0)
+            .chain_update(self.data_hash)
+            .chain_update(chain_id.0);
+
+        if let Some(nonce) = self.nonce {
+            hasher.update(nonce.0);
+        }
+
+        TransactionHash(hasher.finalize())
+    }
+}
+
+#[derive(Default)]
+struct PreV3Hasher {
+    pub prefix: Felt,
+    pub version: TransactionVersion,
+    pub address: ContractAddress,
+    pub entry_point: EntryPoint,
+    pub data_hash: Felt,
+    pub max_fee: Fee,
+    pub nonce_or_class: Option<Felt>,
+    pub casm_hash: Option<CasmHash>,
+}
+
+impl PreV3Hasher {
+    fn hash(self, chain_id: ChainId) -> TransactionHash {
+        let mut hash = PedersenHasher::default()
+            .chain_update(self.prefix)
+            .chain_update(self.version.0)
+            .chain_update(self.address.0)
+            .chain_update(self.entry_point.0)
+            .chain_update(self.data_hash)
+            .chain_update(self.max_fee.0)
+            .chain_update(chain_id.0);
+
+        if let Some(felt) = self.nonce_or_class {
+            hash.update(felt);
+        }
+
+        if let Some(felt) = self.casm_hash {
+            hash.update(felt.0);
+        }
+
+        TransactionHash(hash.finalize())
+    }
+}
+
+/// Provides hashing for V3 transactions.
+struct V3Hasher<'a> {
+    pub prefix: Felt,
+    pub sender_address: ContractAddress,
+    pub nonce: TransactionNonce,
+    pub data_hashes: &'a [Felt],
+    pub tip: Tip,
+    pub paymaster_data: &'a [PaymasterDataElem],
+    pub nonce_data_availability_mode: DataAvailabilityMode,
+    pub fee_data_availability_mode: DataAvailabilityMode,
+    pub resource_bounds: ResourceBounds,
+}
+
+impl V3Hasher<'_> {
+    fn hash(self, chain_id: ChainId) -> TransactionHash {
+        let hasher = PoseidonHasher::default()
+            .chain(self.prefix.into())
+            .chain(TransactionVersion::THREE.0.into())
+            .chain(self.sender_address.0.into())
+            .chain(self.hash_fee_fields().into())
+            .chain(self.hash_paymaster_data().into())
+            .chain(chain_id.0.into())
+            .chain(self.nonce.0.into())
+            .chain(self.pack_data_pack_data_availability().into());
+
+        let hash = self
+            .data_hashes
+            .iter()
+            .fold(hasher, |hasher, &data| hasher.chain(data.into()))
+            .finish();
+
+        TransactionHash(hash.into())
+    }
+
+    fn pack_data_pack_data_availability(&self) -> u64 {
```

Comment:
nit: is this a typo?
```suggestion
    fn pack_data_availability(&self) -> u64 {
```